Source/CarthageKitTests/Fakes/FakeURLSession.swift

```diff
@@ -0,0 +1,41 @@
+import Foundation
+
+final class FakeDownloadTask: NSURLSessionDownloadTask {
+	var started = false
+	override func resume() {
+		self.started = true
+	}
+
+	var cancelled = false
+	override func cancel() {
+		self.cancelled = true
+	}
+}
+
+final class FakeDataTask: NSURLSessionDataTask {
+	var started = false
+	override func resume() {
+		self.started = true
+	}
+
+	var cancelled = false
+	override func cancel() {
+		self.cancelled = true
+	}
+}
+
+final class FakeURLSession: NSURLSession {
```

Comment:
> But I absolutely appreciate you spending time trying to refactor and test this code.

Squee! You're welcome, and I'm happy to do it.

> IMO mocking is a code smell. It works well enough for adding tests to existing code, but it's much better to structure code so that it requires little to no mocking. So I'm not against mocking, but I'd also prefer to find an alternative if possible.

I totally see your point - integration tests are awesome (Speaking of, I have yet another pull request in the works to add a complete integration test suite - using Bash scripts to test the CLI API to make sure that repos are correctly downloaded, built, updated, etc.) - but in a lot of cases (e.g. talking to the network, talking to the file system, talking to other processes in general), you don't want to have to actually do that in the unit test suite. Not only does it make the tests run slower (tests that take more than a few seconds to run are incredibly painful), but it also tends to lead to unused or incorrect code that doesn't get found out because "oh, I won't test that area because it's too painful".

As a related example, one of the things I want to do is essentially mock out calls to xcodebuild in CarthageKitTests. Obviously, we want to still test that frameworks and such get created, but that should be done in an integration test, not a unit test. I think the unit test would be better served to just test that `xcodebuild -project foo.xcodeproj -scheme bar [...]` gets sent to a task service (and also that the actual task service does the right thing). Testing that the framework actually gets built should be left to an integration test.

This way, I can create a new feature that might make a call to something painful to actually run (e.g. support a new way of building projects), test that it makes the right calls to the mocked services, then also run the integration tests to see that everything still works.

In addition, using mocks forces us to use dependency injection, which is a beautiful thing (why get up and go to the fridge and get a beer when I can just ask my robot to get me a beer?). It also forces us to clean up the mess of bare functions that (in mine and other pivots' experience) RX tends to lead to.

I have plenty of things I want to do for Carthage, and more coming along the way. I have an entire desk here at work that is just cards representing things I want to do in Carthage (by the way, how would you like me to communicate this? Individual pull requests for things as I do them might get annoying after a while. I can create an issue for everything, but then the graph of issues (these two refactors depend upon this refactor, this other refactor depends upon two separate refactors, etc.) will get lost. I could also create a single giant issue, and have each individual refactor as a item in a list on that issue. If you have another suggestion, then that's great too. I'd like to be able to get feedback on some proposed refactors before I actually do them.

I really like Carthage, and I want to help improve it and make it easier for others to improve. Thanks for working with me on this.